fcrepo-clients/src/main/java/edu/unc/lib/dl/acl/fcrepo4/InheritedAclFactory.java

```diff
@@ -0,0 +1,201 @@
+		// Checking for unit depth + 1 since the counter will be one past the object checked
+		if (depth == UNIT_PATH_DEPTH + 1) {
+			Set<String> roles = new HashSet<>();
+			roles.add(UserRole.canViewOriginals.getPropertyString());
+			inheritedPrincRoles.put("everyone", roles);
+		}
+
+		return inheritedPrincRoles;
+	}
+
+	private void mergePrincipalRoles(Map<String, Set<String>> basePrincRoles, Map<String,
+			Set<String>> addPrincRoles) {
+		if (basePrincRoles.isEmpty()) {
+			basePrincRoles.putAll(addPrincRoles);
+		} else {
+			// Add to inherited data.  If principals overlap, extra roles are added (not overridden)
+			addPrincRoles.forEach((principal, roles) -> {
+				if (basePrincRoles.containsKey(principal)) {
+					basePrincRoles.get(principal).addAll(roles);
+				} else {
+					basePrincRoles.put(principal, roles);
+				}
+			});
+		}
+	}
+
+	private void revokedPatronRoles(PID pid, Map<String, Set<String>> princRoles,
+			Set<String> patronPrincipals) {
+		// Check each remaining patron principal to see if it still has patron access
+		Iterator<String> patronIt = patronPrincipals.iterator();
+		while (patronIt.hasNext()) {
+			String patronPrinc = patronIt.next();
+			Set<String> roles = princRoles.get(patronPrinc);
+			
+			// Patron access revoked for this principal, so remove it from inherited roles
+			if (!objectPermissionEvaluator.hasPatronAccess(pid, patronPrinc, Permission.viewMetadata)) {
+				patronIt.remove();
+				princRoles.remove(patronPrinc);
+			} else if (!roles.contains(UserRole.canViewMetadata.getPropertyString())
+						&& !objectPermissionEvaluator.hasPatronAccess(pid, patronPrinc, Permission.viewOriginal)) {
+				// Has metadata but not original permission, so downgrading
+				roles.clear();
+				roles.add(UserRole.canViewMetadata.getPropertyString());
+			}
+		}
+	}
+
+	@Override
+	public PatronAccess getPatronAccess(PID target) {
+		List<PID> pidPath = getPidPath(target);
+		if (pidPath.size() < 3) {
+			return null;
+		}
+
+		PatronAccess computedAccess = PatronAccess.parent;
+		for (int i = CONTENT_STARTING_DEPTH; i < pidPath.size(); i++) {
+			PID pathPid = pidPath.get(i);
+
+			PatronAccess access = objectAclFactory.getPatronAccess(pathPid);
+			if (PatronAccess.none.equals(access)) {
+				return PatronAccess.none;
+			}
+
+			if (PatronAccess.authenticated.equals(access)) {
+				computedAccess = access;
+			}
+		}
+
+		return computedAccess;
+	}
+
+	@Override
+	public Date getEmbargoUntil(PID target) {
+		return getPidPath(target).stream()
+				.map(p -> objectAclFactory.getEmbargoUntil(p))
+				.filter(p -> p != null)
+				.max(Comparator.naturalOrder())
+				.orElse(null);
+	}
+
+	@Override
+	public boolean isMarkedForDeletion(PID target) {
+		return getPidPath(target).stream()
+				.anyMatch(p -> objectAclFactory.isMarkedForDeletion(p));
+	}
+
+	private List<PID> getPidPath(PID pid) {
+		List<PID> path = pathFactory.getAncestorPids(pid);
+		path = new ArrayList<>(path.subList(1, path.size()));
```

Comment:
The collections root object, which is the object that exists to be the starting point for the content tree (where you can start following pcdm:hasMember relations).  All content object's should have paths that start with this object otherwise they are orphaned.  Wonder if the object should be renamed though, since it should actually contain Unit objects rather than Collection objects.  I could probably use a constant here too.  I will add a comment here, and need to write a document describing the repo structure